drivers/spi: updated SPI driver interface

haukepetersen opened the pull request:
- made datastructures overridable by CPU implementations
- made some functions void
- updated doxygen
- adapted SPI test
- adapted SPI implementations
- fixed some bugs in spi transfer functions

The main idea behind this API change is to adapt it to a slightly changed paradigm regarding error handling: The concept is to throw errors on initialization, but to not check all parameters everytime afterwards anymore. This leads to more compact and efficient code while not taking much away in terms of error checking.

On top of this, most implementations are implemented in a way, that there return values were quite useless anyway...

This PR is just the first step in 'modernizing' the existing SPI implementations... 

NOTE: This needs some testing on all affected platforms once we agree on the changes...

kaspar030 commented on drivers/include/periph/spi.h:
```diff
@@ -11,13 +11,9 @@
  * @ingroup     driver_periph
  * @brief       Low-level SPI peripheral driver
```
... interface

kaspar030 commented on drivers/include/periph/spi.h:
```diff
@@ -130,104 +143,87 @@ int spi_init_master(spi_t dev, spi_conf_t conf, spi_speed_t speed);
- * @return              -1 on error
  */
-int spi_transfer_byte(spi_t dev, char out, char *in);
+void spi_transfer_byte(spi_t dev, char out, char *in);
```
I would vote for ```char spi_transfer_byte(spi_t dev, char out)```.

haukepetersen replied:
actually, I think I do, too.

jfischer-phytec-iot replied:
I vote for size_t spi_transfer_byte(spi_t dev, uint8_t out, uint8_t *in);

kaspar030 replied:
the result will always be one.

jfischer-phytec-iot replied:
If an octet was transferred. And then .. it would look better if it calls size_t spi_transfer_bytes(spi_t dev, uint8_t *out, uint8_t *in);

haukepetersen replied:
the main concept is actually, that we check for errors during initialization and always expect this function to succeed afterwards... So returning `size_t`, which would always be 1, would be redundant...

kaspar030 replied:
> If an octet was transferred.

spi always transfers one in, one out, as long as the spi device is valid. and that has been checked by ```spi_init_*```.

gebart replied:
In slave mode, where we don't have control over the clock line, the
transmission may fail due to a timeout, I guess. However, I do agree with
Hauke that there is a bit too much error checking than necessary in the
periph drivers.

haukepetersen replied:
@gebart you are right with the slave mode. But the transfer byte/s function are not used there (slave mode is controlled via transmission begin and a ISR callback. Do you have an opinion on moving the `in` value to be returned?

gebart replied:
I would prefer to have in as the return value instead of a pointer. If
nothing else it will allow the compiler to reduce stack usage by placing
the value in a register instead of RAM.

haukepetersen replied:
Seems like we are getting to a majority of votes - I will adapt the PR.

haukepetersen commented:
changed interface slightly, so that `transfer_byte` and `transfer_reg` return the received byte. Also adapted (most) drivers.

gebart commented on cpu/kinetis_common/spi.c:
```diff
@@ -1263,13 +1237,13 @@ int spi_transfer_regs(spi_t dev, uint8_t reg, char *out, char *in, unsigned int
     byte_out = (uint8_t)SPI_IDLE_DATA;
 
-    for (i = 0; i < (int)length; i++) {
+    for (i = 0; i < (int)len; i++) {
```
you can remove the `(int)` cast since both `len` and `i` are `size_t`

haukepetersen replied:
true, will fix.

gebart commented on tests/periph_spi/main.c:
```diff
@@ -242,21 +241,14 @@ int cmd_transfer(int argc, char **argv)
     spi_acquire(spi_dev);
     gpio_clear(spi_cs);
-    res = spi_transfer_bytes(spi_dev, hello, buffer, strlen(hello));
+    spi_transfer_bytes(spi_dev, hello, buffer, strlen(hello));
```
you can replace all of the `strlen(hello)` calls by `sizeof(hello)` if you change the definition of `hello` into `char hello[] = "Hello";`

gebart commented:
I like this improvement.

Could we somehow prepare the SPI interface for hardware CS support too while we are doing API changes?

I don't really know what the best API model would be here, the CS parameter must be somehow conveyed to the SPI driver, either through a function argument or through some global variable (e.g. a driver struct or similar)

haukepetersen commented:
Lets take it step by step I would say. This improvement is basically the base for further work (as not it is possible, to define a 'whatever-you-like-struct' as SPI device identifier on a per-board basis.

With the CS I see a different difficulty: this is how we handle CS inside device drivers. I would hate to see constructs involving ifdefs on weather to call gpio_set/clear manually or not...

haukepetersen commented:
Hm, wait: I might have an idea: We could introduce another parameter for the transfer functions:
```c
void spi_transfer_bytes(spi_t dev, gpio_t cs, ...);
```
Now the spi driver can deicide itself, if it just ignores the given parameter (and uses internal HW CS lines instead). Even both is possible, as you could implement your CPUs spi driver in a way, that it will use some HW CS pin in case the parameter 'GPIO_UNDEF' is given, and use SW CS.

Using an additional parameter has the advantage, that the SPI driver does not need to keep state. Because keeping say an instance struct for each device that is run on a SPI bus, it would lead to redundant data that can be saved this way. What do you say?

gebart commented:
@haukepetersen 
>With the CS I see a different difficulty: this is how we handle CS inside device drivers. I would hate to see constructs involving ifdefs on weather to call gpio_set/clear manually or not...

I completely agree, which is why I would like to open the discussion on how to handle CS lines within the device drivers.

>Hm, wait: I might have an idea: We could introduce another parameter for the transfer functions:
>```c
void spi_transfer_bytes(spi_t dev, gpio_t cs, ...);
```
>Now the spi driver can deicide itself, if it just ignores the given parameter (and uses internal HW CS lines instead). Even both is possible, as you could implement your CPUs spi driver in a way, that it will use some HW CS pin in case the parameter 'GPIO_UNDEF' is given, and use SW CS.

>Using an additional parameter has the advantage, that the SPI driver does not need to keep state. Because keeping say an instance struct for each device that is run on a SPI bus, it would lead to redundant data that can be saved this way. What do you say?

This is similar to what we have been using in Contiki for a while. You need to add one more parameter though, `void spi_transfer_bytes(spi_t dev, gpio_t cs, bool continue, ...);` to tell the spi_transfer_bytes function to keep asserting the CS line after the transfer is completed. Some transactions are made up of multiple transfers while the CS line is held (e.g. write address+read data bytes).

gebart commented:
If we update the SPI API this way it will also be possible to implement DMA transfers for SPI transfers on platforms where that is supported (e.g. Kinetis)

haukepetersen commented:
About the parameter `continue`: I think we might as well find a way around it, maybe by giving gpio_t some defined value to point this out?

For DMA we wouldn't have to change anything, this is already perfectly feasible with todays API (just nobody found the time to implement it for any platform, yet). But if this kind of change would make it easier - even better!

haukepetersen commented:
I think I had one idea more: How about we define a new type for the additional spi transfer function parameter, maybe `spi_cs_t` or something. This type MUST be defined in a way, that `gpio_t` can be cast onto it, but it further allows for additional values. So say the GPIO's are identified by positive integers (`GPIO_0 := 0; GPIO_1 :=1, ..., GPIO_UNDEF := -1`), you could add values as `SPI_CS_HW0 := -2, SPI_CS_HW1 := -3`, ..., SPI_CS_NOCS := INT_MIN` and so on. These values should be of course defined on a CPU basis with whatever values that are the most efficient on a platform. So in a driver device descriptor, you'd have:
```c
struct {
    ...
    spi_cs_t cs;
    ...
} dev_t;
```
which you can initialize by either 
```c
dev_init(..., (spi_cs_t)GPIO(x,y), ...);
```
or
```c
dev_init(..., SPI_CS_HW(x), ...);
```
What do you think?

gebart commented:
@haukepetersen I don't see a way around the `cont` parameter unless we redefine the `cs` parameter to be a bitfield where we can have a "continue" bit and then the GPIO number.

Your idea for the CS parameter and using special values for HW CS is fine by me, that gives us the flexibility to use both GPIO software CS, and HW CS. But it still requires a continue parameter to tell the SPI bus driver whether to de-assert the CS pin after the transaction or not.

PeterKietzmann commented:
@gebart is there a special sensor you are thinking about when proposing the `cont` parameter? I still have problems to see the need of that parameter besides the `spi_transfer_regs` function which sends an address and data during one call, or just resetting the CS line after such a command.

gebart commented:
@PeterKietzmann e.g. https://github.com/RIOT-OS/RIOT/blob/master/drivers/ng_at86rf2xx/ng_at86rf2xx_internal.c#L68

gebart commented:
it is possible to find many more examples if you look for it, both in implemented drivers and future device drivers.

PeterKietzmann commented:
@gebart thanks! It's been a while writing such a driver. My first idea was to use the transfer_regs function instead of calling transfer_re and transfer_bytes. Anyway, this could lead to ugly results (if that's at least possible). Guess I have to think about it a bit.

gebart commented:
@PeterKietzmann Some devices reset their internal communication state (e.g an address pointer to an SRAM location), when the CS is de-asserted, so it will have to be handled in a single `_regs` call, which will lead to the implementation of some temporary transfer buffers on the MCU side for the at86rf2xx SRAM write case, and I am not sure it is possible to implement SRAM read with the current API in a single call.

@PeterKietzmann Do you agree that a continue parameter is a good addition to the API after seeing these examples?

PeterKietzmann commented:
I agree that it is a good parameter when introducing the HW CS capability. Also I see that HW CS seems to be needed as this topic occurs periodically. I just don't like the idea of having this parameter in the API but using the external CS line without the need of it. That's why I have to think about it a while or become friends with that idea :-)

gebart commented:
Read Hauke's reply above on moving the gpio pin to a function argument of
the transfer function. In my view this is the best solution, even for
software CS. This means simpler device drivers, regardless of what platform
we are running on.
On Jun 29, 2015 11:28 AM, "Peter Kietzmann" <notifications@github.com>
wrote:

> I agree that it is a good parameter when introducing the HW CS capability.
> Also I see that HW CS seems to be needed as this topic occurs periodically.
> I just don't like the idea of having this parameter in the API but using
> the external CS line without the need of it. That's why I have to think
> about it a while or become friends with that idea :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3216#issuecomment-116566596>.
>

PeterKietzmann commented:
Ok, that would force us to use a `cont` parameter in either ways, right?

gebart commented:
Yes
On Jun 29, 2015 11:48 AM, "Peter Kietzmann" <notifications@github.com>
wrote:

> Ok, that would force us to use a cont parameter in either ways, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3216#issuecomment-116584765>.
>

haukepetersen commented:
@gebart: You are right, the `continue` parameter is needed (I think I overlooked that one before...). So we are at
```c
char spi_transfer_byte(spi_t spi, spi_cs_t cs, bool continue, char data);
void spi_transfer_bytes(spi_t spi, spi_cs_t cs, bool continue, char *data, size_t len);
```
This goes together with the assumption, that chip select is taken care of internally inside the transfer functions. 

What do we do about locking? I think for simplicity it makes sense to keep it externally to the transfer functions...

gebart commented:
I agree that we should leave locking outside the transfer function for now. There may be cases where it is not necessary or desirable to do locking (the kw2x radio for example)

haukepetersen commented:
true. I will adjust this PR to the discussed changes, hopefully I get to it tomorrow.

PeterKietzmann commented:
+1 for leaving the locking outside

haukepetersen commented:
One thing I just thought about: Moving the CS code inside the SPI driver would also mean, that we always assert that the CS active state is low. In the proposed design, we so far have no means to define the active state of the CS line. With the old implementation, this was not an issue... 

I would suggest we just go ahead, as I think 99.9% of the devices are using low-active CS signals, anyway. Or does somebody have an example proving me wrong?

haukepetersen commented:
Question2: So far we have the `spi_transfer_reg(s)` convenience functions in the driver interface. By looking at their implementations and the device drivers, I actually tend to drop these. This would make the SPI interface slimmer and save two implementations per CPU, that look more or less identical on all platforms. The device driver code in turn does not become more complicated by falling back to the `spi_transfer_byte(s)` functions -> in the end we save a function call layer and save some bytes of flash.

Any opinion on dropping these two functions?

gebart commented:
@haukepetersen I agree regarding `spi_transfer_reg(s)`, either remove it or move it to the spi header and make it inline wrappers for `spi_transfer_byte(s)`.

gebart commented:
@haukepetersen HW CS already needs the CS polarity defined at initialization time, perhaps when someone show such a device that needs active high CS we can add an inversion/polarity flag to `gpio_init`?

haukepetersen commented:
Updated the SPI interface header file. TODO: adjust implementations and remove transfer_reg functions

haukepetersen commented:
@gebart: The problem I see, is that by making the CS line implicit by adding it as paramter to the transfer functions, we decoupled it from the init function... 

I also notices a second problem we just run into: How to we handle the initialization of the CS line? As we now made internal CS lines possible, we can not really call `init_gpio()` inside a device driver initialization section anymore - as we might be using internal CS lines, which have a different init sequence. Now thinking out loud: how about we expand the SPI interface with CS handling funcions:
```c
int spi_cs_init(spi_t dev, spi_cs_t cs, spi_cs_pol_t pol);
static void spi_cs_on(spi_t dev, spi_cs_t);
static void spi_cs_off(spi_t dev, cpi_cs_t);
```
In the simplest case, init would call `gpio_init`, and `on/off` would call `gpio_clear/set`. In the case the first parameter would not be needed. But I think in cases for HW CS we might need it?! Furthermore the on/off functions should probably just be internal functions for the SPI driver implementation, as they are only used internal...

gebart commented:
Good point.
You are right that we need the `spi_t` parameter to cs init, at least on Kinetis it is necessary.

I don't like the names "_on" and "_off", since that might lead us to think in terms of on as in digital high. How about something else which goes more with the logical state "assert"/"deassert", but assert in software context makes me think of debugging checks. Suggestions?

kaspar030 commented:
> Any opinion on dropping these two functions?

+1

haukepetersen commented:
> How about something else which goes more with the logical state "assert"/"deassert",

+1

PeterKietzmann commented:
- Having a `spi_cs_init()` function is not very pretty but needed I guess
- Even if I don't know a device that needs a CS high signal to be activated I think it is reasonable to have this option in the interface
- I liked the `spi_transfer_reg(s)` but I also think that we can drop them
- who about "select"/"deselect" instead of "assert"/"deassert"?

haukepetersen commented:
About the naming of the spi_cs_(assert|set|select) functions: I think we can stop the discussion here, as these functions are purely interal of the SPI driver and so everyone implementing an SPI driver can just chose names they see fit.

gebart commented:
@haukepetersen Will you update this PR with the `spi_cs_` functions?

@PeterKietzmann good suggestions.

haukepetersen commented:
Yes, I try to get it done before I go into holiday tomorrow evening...

haukepetersen commented:
Updated the SPI interface (but not the implementations/drivers, yet):

- Introduced the `spi_init_cs()` function
- added some data-structures concerning chip select
- adapted the `spi_init_slave()` function quite a bit to work with internal CS handling
- removed the now not needed 'spi_transmission_begin()` function

I think the slave mode is much more straight forward this way. @gebart: Does the slave mode make sense when using HW CS lines? Never worked with HW CS in slave mode...

I am on holyday for the next week and I won't probably be able to adapt all the implementations to the latest interface changes before I leave, so I try to update this PR first thing when I come back...

gebart commented:
​The hardware CS in slave mode on kinetis is limited to CS line 0, which will let the hardware know if it should send anything on the SPI bus when it is being clocked, I'm not sure how you would do slave mode without hardware CS handling.

I don't have a use case for SPI slave mode so I haven't had any motivation to spend any effort on getting it to work though.

gebart commented:
needs rebase

haukepetersen commented:
working on it... ETA: next week.

PeterKietzmann commented:
needs rebase again

haukepetersen commented:
I know, I know :-)

gebart commented:
ping @haukepetersen

haukepetersen commented:
got the ping, will try to get to it...

haukepetersen commented:
rebased, though the PR needs some love...

haukepetersen commented:
did some more changes (everything pretty much WIP):
- changed the interface to use `uint8_t` instead of `char`
- remodeled the test application to demonstrate the usage of the interface

gebart commented:
needs rebase

haukepetersen commented:
not only that :-) Can't promise when I get to it, hopefully soon!

OlegHahm commented:
Probably not before Monday, right?

haukepetersen commented:
definitely not...

authmillenon commented:
How about Tuesday ;-)?

haukepetersen commented:
Some Tuesday. It is still on my list, but it seems there is more important stuff coming up all the time...

kYc0o commented:
any news on this PR?

haukepetersen commented:
still trying to rework the concept for shared buses. I think I found a suitable solution, but need to investigate some more first. Expect to have something for the next release at the latest...

gebart commented:
should we close this in favour of #4780?

haukepetersen commented:
Makes sense. I left this open for some readon, but I can't remember it anymore, so I can't have been important :-)
